nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -159,23 +203,37 @@ public void checkTypeParameterNullnessForAssignability(Tree tree) {
     if (rhsTree == null || rhsTree.getKind().equals(Tree.Kind.NULL_LITERAL)) {
       return;
     }
-    Type lhsType = ASTHelpers.getType(lhsTree);
-    Type rhsType = ASTHelpers.getType(rhsTree);
-    // For NewClassTrees with annotated type parameters, javac does not preserve the annotations in
-    // its computed type for the expression.  As a workaround, we construct a replacement Type
-    // object with the appropriate annotations.
-    if (rhsTree instanceof NewClassTree
-        && ((NewClassTree) rhsTree).getIdentifier() instanceof ParameterizedTypeTree) {
-      ParameterizedTypeTree paramTypedTree =
-          (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier();
-      if (paramTypedTree.getTypeArguments().isEmpty()) {
-        // no explicit type parameters
-        return;
+    Type lhsType = getTreeType(lhsTree);
+    Type rhsType = getTreeType(rhsTree);
+
+    if (lhsType instanceof Type.ClassType && rhsType instanceof Type.ClassType) {
+      boolean isAssignmentValid =
+          compareNullabilityAnnotations((Type.ClassType) lhsType, (Type.ClassType) rhsType);
+      if (!isAssignmentValid) {
+        reportInvalidAssignmentInstantiationError(tree, lhsType, rhsType, state, analysis);
       }
-      rhsType = typeWithPreservedAnnotations(paramTypedTree);
     }
-    if (lhsType instanceof Type.ClassType && rhsType instanceof Type.ClassType) {
-      compareNullabilityAnnotations((Type.ClassType) lhsType, (Type.ClassType) rhsType, tree);
+  }
+
+  public void checkTypeParameterNullnessForFunctionReturnType(
+      ExpressionTree retExpr, Symbol.MethodSymbol methodSymbol) {
+    if (!config.isJSpecifyMode()) {
+      return;
+    }
+
+    Type methodType = methodSymbol.getReturnType();
```

Comment:
`methodType` is a confusing name for this. I would suggest `formalReturnType` or `returnTypeFromDeclaration`?